### PR TITLE
Conecting frontend and backend

### PR DESCRIPTION
**Pull request by SoyOscarRH**
* Finally using the fudge ratio servlet 
* Better value to save for activity
* Show average fudge ratio

loco
Screen Shot 2020-04-12 at 2 53 33 AM

### CODE REVIEW

File: fudge-ratio/src/main/webapp/code/utils.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+const get = id => document.getElementById(id);
+
```
Comment (yaaowusu00) [inline fudge-ratio/src/main/webapp/code/utils.js]: this is clever, I never thought about doing something like this

File: fudge-ratio/src/main/webapp/code/time.js [hunk 2]
```diff
@@ -1,52 +1,46 @@
...
-    this.seconds++;
-    if (this.seconds == 60) {
-      this.minutes++;
-      this.seconds = 0;
-    }
-    if (this.minutes == 60) {
-      this.hours = (this.hours + 1) % 24;
-      this.minutes = 0;
-    }
+  getHuman() {
```
Comment (yaaowusu00) [inline fudge-ratio/src/main/webapp/code/time.js]: I might just be confused, but why is this called getHuman?

File: fudge-ratio/src/main/java/com/google/sps/servlets/FudgeRatio.java [hunk 3]
```diff
@@ -1,21 +1,34 @@
...
+		final int expected = Integer.parseInt(request.getParameter("expected"));
 
-    @Override
-    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    
-    }
+		final Gson gson = new Gson();
+		final Map < String,
+		Double > result = new HashMap < String,
+		Double > ();
```
Comment (yaaowusu00) [inline fudge-ratio/src/main/java/com/google/sps/servlets/FudgeRatio.java]: This seems like it should be on one or two lines right? It seems like the lines are split in a weird place.

File: fudge-ratio/src/main/java/com/google/sps/servlets/FudgeRatio.java [hunk 4]
```diff
@@ -6,24 +6,26 @@
...
+
+		final String start = data.get("start").toString();
+		final String end = data.get("end").toString();
+		final double expected = (double) data.get("expected");
 
 		final Gson gson = new Gson();
-		final Map < String,
-		Double > result = new HashMap < String,
-		Double > ();
+		final Map <String,Double> result = new HashMap <String,Double> ();
```
Comment (yaaowusu00) [inline fudge-ratio/src/main/java/com/google/sps/servlets/FudgeRatio.java:28]: I see you fixed it here. Ignore my comment on the last commit.

File: fudge-ratio/src/main/webapp/code/chooseActivities.js [hunk 5]
```diff
@@ -34,21 +34,21 @@ const appendActivities = async () => {
...
+    appendActivities();
 
     const submitButton = get("submit");
     const options = get("options");
     options.onchange = () => (submitButton.disabled = options.value == "");
 
     submitButton.onclick = () => {
       if (submitButton.disabled) return;
-      localStorage.setItem(KEYS.INDEX_KEY, options.selectedIndex);
+      localStorage.setItem(KEYS.ACTIVITY_KEY, options.value);
```
Comment (yaaowusu00) [inline fudge-ratio/src/main/webapp/code/chooseActivities.js:51]: I really like how simple this is now

File: fudge-ratio/src/main/webapp/code/mainScreen.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      const response = await fetch("/fudgeRatio", {
+        method: "POST",
+        headers: { Accept: "application/json", "Content-Type": "application/json;charset=UTF-8" },
+        body: JSON.stringify({ start, end, expected }),
+      });
+
+      const { ratio } = await response.json();
+      get("result").textContent = `Next time think about allocating ${ratio.toFixed(2)}x 
+      ${ratio > 1 ? "more" : "less"}
+      for ${activity}`;
```
Comment (yaaowusu00) [inline fudge-ratio/src/main/webapp/code/mainScreen.js:79]: This is me being very picky, but the reason I had a method before called format task was just so that when we display  "Next time think about allocating ... "  the sentence is is just a little nicer (theres picture of that old function below as a reminder). For example, if the person selects "English"  as their task ( under the homework category) then the final sentence would say "think about allocating .. for english **homework** . Another small thing is that I think it would be more effective to actually display the recommended **time** in the sentence because for a lot of people, its hard to conceptualize the ratio, especially if it is a decimal. In your example gif, I think some people would have a hard time with really visualizing  .62 of their original time (and remembering this in the future to help them change their behavior) without going off on their own and using a calculator to convert it to actual minutes/hours. Let me know what you think about this!
Screen Shot 2020-04-16 at 10 58 02 AM

**Review by yaaowusu00 — CHANGES_REQUESTED** [review]
I left a comment about two really small things that we should probably decide on as a team. Otherwise I think this was great. I love the way you simplified local storage!

File: fudge-ratio/src/main/webapp/code/utils.js (hunk 1 above)
Comment (SoyOscarRH) [reply fudge-ratio/src/main/webapp/code/utils.js]: It is not super professional, but it get the job done and eliminate the mental burden of reading a really long method name

File: fudge-ratio/src/main/webapp/code/time.js (hunk 2 above)
Comment (SoyOscarRH) [reply fudge-ratio/src/main/webapp/code/time.js]: This class store time as a number of seconds, so if I tell you, the event happened 3231 seconds ago, is this useful for you? humans think in terms of minutes and hours too, so if the class tell you, hey, the event happened 2 hours, and 34 minutes ago. That is a lot more useful, but the information is the same.

So getHumans, get the information, but in a human understandable way.
That was my thinking, but again, ``naming is hard``

Also, I deleted the method later as it was only used for getText

File: fudge-ratio/src/main/java/com/google/sps/servlets/FudgeRatio.java (hunk 4 above)
Comment (SoyOscarRH) [reply fudge-ratio/src/main/java/com/google/sps/servlets/FudgeRatio.java:28]: Yeah, turn out I closed the prettier commit was never accepted to without a config file I think prettier  made a couple of mistakes.

File: fudge-ratio/src/main/webapp/code/mainScreen.js (hunk 6 above)
Comment (SoyOscarRH) [reply fudge-ratio/src/main/webapp/code/mainScreen.js:79]: This is a really GREAT comment, and I completely agree with you.
https://gph.is/2aUc9VT

The general idea is that this need improvement, and I am completely  on your side, to be totally honest, I refactored the excepted time, the activity selector and then the main screen, so I was to tired to create a new "screen" for the final result, but I think the final result deserve one, so my idea was that when your finish the time, a transition occurs where you have a big message with the result, maybe some analytics but the most important part was that message "PLEASE ALLOCATE ..." and then maybe a couple of buttons to share the app or start a new timer.

In part I left that undo because I was to tired but also because I remembered that @gilemos, wanted to to do some frontend, so I thought that this was perfect for a task this week, create a final screen with the correct message (as you said in a more readable way, or Human way, maybe is time to get back to life the ``getHuman`` 😅), I think i left a comment in my last PR mentioning this exact situation.

So TLDR; I agree with the 2 main points, we should use your idea to create a better message and give that message in a more readable way.

I thought that @gilemos could do that, but if she is too busy maybe you or me could do it, I see no problem doing it myself :3

Screen Shot 2020-04-16 at 4 53 35 PM

Comment (SoyOscarRH) [reply fudge-ratio/src/main/webapp/code/mainScreen.js:79]: Thanks @yaaowusu00 for taking the time for really checking the changes and giving your honest opinions

File: fudge-ratio/src/main/webapp/code/chooseActivities.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  </select>
+  <button disabled id="submit">Add activity</button>
+</div>
+  `;
+
+import { get, parent, createFromText } from "./utils.js";
+import { KEYS } from "./save.js";
+
+const getActivities = async () => {
+  const response = await fetch("/data");
```
Comment (gilemos) [inline fudge-ratio/src/main/webapp/code/chooseActivities.js:19]: Does that still work (since we do not have a /data anymore) ?

File: fudge-ratio/src/main/webapp/code/mainScreen.js (hunk 6 above)
Comment (yaaowusu00) [reply fudge-ratio/src/main/webapp/code/mainScreen.js:79]: I am really busy with final exams so I'm not sure if I can get to it 😞

File: fudge-ratio/src/main/webapp/code/chooseActivities.js (hunk 7 above)
Comment (fancast) [reply fudge-ratio/src/main/webapp/code/chooseActivities.js:19]: I believe it should still work since the Activity servlet still exists with that endpoint.

File: fudge-ratio/src/main/webapp/code/expectedTime.js [hunk 8]
```diff
@@ -1,96 +1,64 @@
...
-buttonMinusSecond.addEventListener("click", () => {
-  if (durationSec > 0) {
-    durationSec -= 1;
-  }
-  durationText.textContent = padd(durationHour) + ":" + padd(durationMin) + ":" + padd(durationSec);
-});
+    parent.appendChild(createFromText(expectedTimeText));
+    const text = get("timer-display");
+    const updateAndShow = time => () => (text.textContent = timer.addSeconds(time).getText());
+    get("add-second").onclick = updateAndShow(1);
```
Comment (fancast) [inline fudge-ratio/src/main/webapp/code/expectedTime.js:36]: Much simpler

**Review by yaaowusu00 — APPROVED** [review]
I don't know if any of us will get to adding anything extra so I'm going to approve this so its not blocked from merging
